plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+      }
+      processMetadata(metadataMavenPath(groupId, artifactId, null), metadataBuilder.onExitArtifactId());
+    }
+
+    /**
+     * Verifies and may fix/create the broken/non-existent Maven hashes (.sha1/.md5 files).
+     */
+    private void mayUpdateChecksum(final Asset asset, final MavenPath mavenPath, final HashType hashType) {
+      final AttributesMap checksums = asset.attributes().child(StorageFacet.P_CHECKSUM);
+      checkState(checksums != null, "checksums");
+      final String assetChecksum = (String) checksums.get(hashType.getHashAlgorithm().name());
+      checkState(!Strings.isNullOrEmpty(assetChecksum), "assetChecksum");
+      final MavenPath checksumPath = mavenPath.hash(hashType);
+      try {
+        final Content content = mavenFacet.get(checksumPath);
+        if (content != null) {
+          try (InputStream is = content.openInputStream()) {
+            final String mavenChecksum = DigestExtractor.extract(is);
+            if (Objects.equals(assetChecksum, mavenChecksum)) {
+              return; // all is OK: exists and matches
+            }
+          }
+        }
+      }
+      catch (IOException e) {
+        // TODO: ignore that underlying storage puke on read?
+        log.warn("Error reading {}", checksumPath, e);
+      }
+      // we need to generate/write it
+      try {
+        final StringPayload mavenChecksum = new StringPayload(assetChecksum, Maven2Format.CHECKSUM_CONTENT_TYPE);
+        mavenFacet.put(checksumPath, mavenChecksum);
+      }
+      catch (IOException e) {
+        log.warn("Error writing {}", checksumPath, e);
+        throw Throwables.propagate(e);
+      }
+    }
+
+    /**
+     * Assembles {@link MavenPath} for repository metadata out of groupId, artifactId and baseVersion.
+     */
+    private MavenPath metadataMavenPath(final String groupId,
+                                        @Nullable final String artifactId,
+                                        @Nullable final String baseVersion)
+    {
+      final StringBuilder sb = new StringBuilder("/");
+      sb.append(groupId.replace('.', '/'));
+      if (artifactId != null) {
+        sb.append("/").append(artifactId);
+        if (baseVersion != null) {
+          sb.append("/").append(baseVersion);
+        }
+      }
+      sb.append("/").append(Maven2Format.METADATA_FILENAME);
+      return mavenPathParser.parsePath(sb.toString());
+    }
+
+    /**
+     * Reads and parses Maven POM.
+     */
+    @Nullable
+    private Document getModel(final MavenPath mavenPath) {
+      // sanity checks: is artifact and extension is "pom", only possibility for maven POM currently
+      checkArgument(mavenPath.getCoordinates() != null);
+      checkArgument(Objects.equals(mavenPath.getCoordinates().getExtension(), "pom"));
+      try {
+        final Content pomContent = mavenFacet.get(mavenPath);
+        if (pomContent != null) {
```

Comment:
IMHO, it is not, code must be robust, especially if we do distributed db. As I understand Orient docos, it might be gone during "walking", similarly to this case:
![It's gone]
(http://www.silverdoctors.com/wp-content/uploads/2014/02/bitcoin-gone.jpg)